pandas/core/algorithms.py

```diff
@@ -142,6 +142,71 @@ def isin(comps, values):
     return f(comps, values)
 
 
+def safe_sort(values, labels=None, na_sentinel=-1, assume_unique=False):
+    """
+    Sort ``values`` and reorder corresponding ``labels``.
+    ``values`` should be unique if ``labels`` is not None.
+    Safe for use with mixed types (int, str), orders ints before strs.
+
+    Parameters
+    ----------
+    values : ndarray (1-d)
+        Sequence; must be unique if ``labels`` is not None.
+    labels : ndarray (1-d)
+        Indices to ``values``
+    na_sentinel : int, default -1
+        Value in ``labels`` to mark "not found".
+        Ignored when ``labels`` is None.
+    assume_unique : bool, default False
+        When True, doesn't check for uniqness of ``values``.
+        Ignored when ``labels`` is None.
+
+    Returns
+    -------
+    ordered : ndarray
+        Sorted ``values``
+    new_labels : ndarray
+        Reordered ``labels``; returned when ``labels`` is not None
+    """
+    def sort_mixed(values):
+        # order ints before strings, safe in py3
+        str_pos = np.array([isinstance(x, string_types) for x in values],
+                           dtype=bool)
+        nums = np.sort(values[~str_pos])
+        strs = np.sort(values[str_pos])
+        return com._ensure_object(np.concatenate([nums, strs]))
+
+    sorter = None
+    try:
+        sorter = values.argsort()
+        ordered = values.take(sorter)
+    except:
+        # unorderable in py3 if mixed str/int
+        ordered = sort_mixed(values)
```

Comment:
this would be fine. the inference routines are pretty fast as they short circuit.